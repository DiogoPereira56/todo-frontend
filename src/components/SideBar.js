import styled from 'styled-components'
import '../fonts.css'
import menu from '../imgs/menu.png'


const Wrapper = styled.div`
    display: flex;
    position: relative;
`
const SideBar2 = styled.div`
    width: 290px;
    height: 100%;
    background: #eeeeee;
    padding: 30px 0px;
    position: fixed;
`
const H2 = styled.h2`
    text-align: left;
    margin-bottom: 20px;
    margin-left: 10px;
`
const Img = styled.img`
    width: 40px;
    
    &:hover{
        background: rgb(248, 248, 248);
    }
`
const Li = styled.li`
    padding: 15px;
    border-bottom: 1px solid #bdb8d7;
    border-bottom: 1px solid rgba(0,0,0,0.05);
    border-top: 1px solid rgba(255,255,255,0.05);

    &:hover{
        background: rgb(248, 248, 248);
    }
`
const P = styled.p`
    color: black;
    display: block;
`
const Input = styled.input`
    border: none !important;
    background: #eeeeee;
    font-size: 1.4rem;
    width: 100%;
    padding: 6px 6px 8px 6px;

    &:hover{
        background: rgb(248, 248, 248);
    }
`
const Ul = styled.ul`
    margin-bottom: 20px;
`

const SideBar = () => {
    return (
        <Wrapper>
            <SideBar2>
                <H2><Img src={menu} alt="" /></H2>
                <Ul>
                    <Li><P>My day</P></Li>
                    <Li><P>Important</P></Li>
                    <Li><P>Planned</P></Li>
                    <Li><P>Assigned to you</P></Li>
                    <Li><P>Flagged email</P></Li>
                    <Li><P>Tasks</P></Li>
                </Ul> 
                <Input placeholder="+ New List"></Input>
            </SideBar2>
        </Wrapper>
        );
    }
    
    export default SideBar;