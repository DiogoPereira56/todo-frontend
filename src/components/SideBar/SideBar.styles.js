import styled from 'styled-components';

const Wrapper = styled.div`
    display: flex;
    position: relative;
    margin-right: 290px;
`;
const SideBar2 = styled.div`
    width: 290px;
    height: 100%;
    background: #eeeeee;
    padding: 30px 0px;
    position: fixed;
`;
const H2 = styled.h2`
    text-align: left;
    margin-bottom: 20px;
    margin-left: 10px;
`;
const Img = styled.img`
    width: 35px;

    &:hover {
        background: rgb(248, 248, 248);
    }
`;
const Li = styled.li`
    padding: 10px;
    border-bottom: 1px solid #bdb8d7;
    border-bottom: 1px solid rgba(0, 0, 0, 0.05);
    border-top: 1px solid rgba(255, 255, 255, 0.05);
    cursor: pointer;

    &:hover {
        background: rgb(248, 248, 248);
    }
`;
const P = styled.p`
    color: black;
    display: block;
    font-size: 0.8rem;
`;
const Input = styled.input`
    border: none !important;
    background: #eeeeee;
    font-size: 1rem;
    width: 100%;
    padding: 6px 6px 8px 6px;

    &:hover {
        background: rgb(248, 248, 248);
    }
`;
const Ul = styled.ul`
    margin-bottom: 20px;
`;
const PaginationWrapper = styled.div`
    display: flex;
    margin: 7px;
`;

const Paginnation = styled.div`
    margin-left: 3px;
    margin-right: 3px;
    width: 20px;
    text-align: center;
    border: solid black 1px;

    cursor: pointer;

    &:hover {
        background: rgb(248, 248, 248);
    }
`;

export { Wrapper, SideBar2, H2, Img, Li, P, Input, Ul, PaginationWrapper, Paginnation };
