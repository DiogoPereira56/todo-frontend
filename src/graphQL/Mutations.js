import { gql } from '@apollo/client';

//User login and resgistration
export const LOGIN_USER_MUTATION = gql`
    mutation clientLogin($email: String!, $password: String!) {
        login(email: $email, password: $password)
    }
`;

export const REGISTER_USER_MUTATION = gql`
    mutation clientRegister($email: String!, $name: String!, $password: String!) {
        register(name: $name, email: $email, password: $password)
    }
`;

export const CLIENT_LISTS_MUTATION = gql`
    mutation getClient($limit: Float!, $offset: Float!) {
        getClientInformations {
            name
            idClient
            list(limit: $limit, offset: $offset) {
                idList
                idClient
                listName
            }
        }
    }
`;

//List actions
export const NEW_LIST_MUTATION = gql`
    mutation addNewList($listName: String!) {
        addList(listName: $listName)
    }
`;

export const DELETE_LIST_MUTATION = gql`
    mutation($idList: Float!, $idClient: Float!) {
        removeList(idList: $idList, idClient: $idClient)
    }
`;

export const RENAME_LIST_MUTATION = gql`
    mutation newList(
        $idList: Float!
        $title: String!
        $idClient: Float!
        $limit: Float!
        $offset: Float!
        $orderByTitle: Boolean!
        $order: String!
    ) {
        updateList(idList: $idList, title: $title, idClient: $idClient) {
            idList
            listName
            taskss(limit: $limit, offset: $offset, orderByTitle: $orderByTitle, order: $order) {
                idTask
                title
                complete
                description
            }
        }
    }
`;

export const LIST_INFO_MUTATION = gql`
    mutation getListInfo(
        $idList: Float!
        $idClient: Float!
        $limit: Float!
        $offset: Float!
        $orderByTitle: Boolean!
        $order: String!
    ) {
        getList(idList: $idList, idClient: $idClient) {
            idList
            idClient
            listName
            taskss(limit: $limit, offset: $offset, orderByTitle: $orderByTitle, order: $order) {
                idTask
                title
                complete
                description
            }
        }
    }
`;

export const GET_CLIENT_TOTAL_LISTS = gql`
    mutation {
        getClientTotalLists
    }
`;

//Task Actions
export const NEW_TASK_MUTATION = gql`
    mutation newTask($title: String!, $idList: Float!, $idClient: Float!) {
        addTask(title: $title, idList: $idList, idClient: $idClient) {
            idTask
            idList
            title
            complete
            description
        }
    }
`;

export const DELETE_TASK_MUTATION = gql`
    mutation($idTask: Float!, $idClient: Float!) {
        removeTask(idTask: $idTask, idClient: $idClient)
    }
`;

export const UPDATE_TASK_DESCRIPTION_MUTATION = gql`
    mutation($idTask: Float!, $description: String!, $idClient: Float!) {
        updateTaskDescription(idTask: $idTask, description: $description, idClient: $idClient) {
            idTask
            title
            complete
            description
        }
    }
`;

export const UPDATE_TASK_TITLE_MUTATION = gql`
    mutation($idTask: Float!, $title: String!, $idClient: Float!) {
        updateTaskTitle(idTask: $idTask, title: $title, idClient: $idClient) {
            idTask
            title
            complete
            description
        }
    }
`;

export const UPDATE_TASK_COMPLETION_MUTATION = gql`
    mutation($idTask: Float!, $complete: Boolean!, $idClient: Float!) {
        updateTaskCompletion(idTask: $idTask, complete: $complete, idClient: $idClient) {
            idTask
            idList
            title
            complete
            description
        }
    }
`;

export const LIST_TOTAL_TASK_MUTATION = gql`
    mutation listTotalTasks($idList: Float!, $idClient: Float!) {
        getListsTotalTasks(idList: $idList, idClient: $idClient)
    }
`;

export const ALL_CLIENT_TASKS_MUTATION = gql`
    mutation($limit: Float!, $offset: Float!, $idClient: Float!, $orderByTitle: Boolean!, $order: String!) {
        getAllTasks(
            limit: $limit
            offset: $offset
            idClient: $idClient
            orderByTitle: $orderByTitle
            order: $order
        ) {
            idTask
            idList
            title
            complete
            description
        }
    }
`;
export const SEARCHED_TASKS_MUTATION = gql`
    mutation(
        $limit: Float!
        $offset: Float!
        $idClient: Float!
        $orderByTitle: Boolean!
        $search: String!
        $order: String!
    ) {
        getSearchedTasks(
            limit: $limit
            offset: $offset
            idClient: $idClient
            orderByTitle: $orderByTitle
            search: $search
            order: $order
        ) {
            idTask
            idList
            title
            complete
            description
        }
    }
`;
export const TOTAL_SEARCHED_TASKS_MUTATION = gql`
    mutation($idClient: Float!, $search: String!) {
        getTotalSearchedTasks(idClient: $idClient, search: $search)
    }
`;
