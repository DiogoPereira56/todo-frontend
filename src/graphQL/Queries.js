import {gql} from '@apollo/client'


export const DECODED_TOKEN = gql`
  {
    getDecodedToken{
      id
      name
    }
  }
`

export const LOG_OUT = gql`
  {
    removeToken
  }
`

export const GET_ALL_LISTS = gql`
  {
    lists{
      idList
      idClient
      listName
    }
  }
`

export const GET_CLIENT_INFORMATION = gql`
{
  getClientInformation{
    name
    lists{
      idList
      listName
      tasks{
        idTask
        title
        complete
        description
      }
    }
  }
}
`
