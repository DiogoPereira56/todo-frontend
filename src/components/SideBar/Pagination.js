import { PropTypes } from 'prop-types'
import { PaginationWrapper, Paginnation } from './SideBar.styles'

const Pagination = ({ listsPerPage, totalLists, setCurrentPage }) => {
    const pageNumbers = [];

    for (let i = 1; i <= Math.ceil(totalLists / listsPerPage); i++){
        pageNumbers.push(i);
    }

    return (
        <nav>
            <PaginationWrapper>
                {pageNumbers.map(number => (
                    <Paginnation key={number} onClick={() => setCurrentPage(number)} >
                        {number}
                    </Paginnation>
                ))}
            </PaginationWrapper>
        </nav>
    )

}

Pagination.propTypes = {
    listsPerPage: PropTypes.number,
    totalLists: PropTypes.number,
    setCurrentPage: PropTypes.func
  };

export default Pagination;