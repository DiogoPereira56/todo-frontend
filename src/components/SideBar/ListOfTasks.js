import { useMutation } from '@apollo/client';
import { PropTypes } from 'prop-types';
import { LIST_INFO_MUTATION } from '../../graphQL/Mutations';
import { Li, P } from './SideBar.styles';

const ListOfTasks = ({
    lists,
    setActiveList,
    setChangeLayout,
    setRename,
    setShowOptions,
    setShowAllTasks,
    orderByTitle,
    setSearchIsActive,
    order,
    setCurrentTaskPage,
    loadListInfo,
}) => {
    const [getListTasks] = useMutation(LIST_INFO_MUTATION);

    function changeActiveList(list) {
        //console.log(list.idList, list.idClient);
        getListTasks({
            variables: {
                idList: list.idList,
                idClient: list.idClient,
                limit: 13,
                offset: 0,
                orderByTitle: orderByTitle,
                order: order,
            },
        }).then((data) => {
            if (data.data) {
                setActiveList(data.data.getList);
                //console.log(data.data.getList);
            }
        });

        loadListInfo({
            variables: {
                idList: list.idList,
                idClient: list.idClient,
                limit: 13,
                offset: 0,
                orderByTitle: orderByTitle,
                order: order,
            },
        });
    }

    const makeActiveList = (list) => {
        //setActiveList(list);
        changeActiveList(list);
        setShowAllTasks(false);
        setChangeLayout(false);
        setRename(false);
        setShowOptions(false);
        setSearchIsActive(false);
        setCurrentTaskPage(1);
    };

    return (
        <div>
            {lists.map((list) => (
                <Li key={list.idList} onClick={() => makeActiveList(list)}>
                    <P>{list.listName}</P>
                </Li>
            ))}
        </div>
    );
};

ListOfTasks.propTypes = {
    lists: PropTypes.array.isRequired,
    setActiveList: PropTypes.func,
    setChangeLayout: PropTypes.func,
    setRename: PropTypes.func,
    setShowOptions: PropTypes.func,
    setShowAllTasks: PropTypes.func,
    orderByTitle: PropTypes.bool,
    setSearchIsActive: PropTypes.func,
    order: PropTypes.string,
    setCurrentTaskPage: PropTypes.func,
    loadListInfo: PropTypes.func,
};

export default ListOfTasks;
