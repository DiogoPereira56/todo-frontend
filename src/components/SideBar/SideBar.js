import {Wrapper, SideBar2, H2, Img, Li, P, Input, Ul} from './SideBar.styles'
import '../../fonts.css'
import menu from '../../imgs/menu.png'
import { useMutation, useQuery } from '@apollo/client';
import { GET_ALL_LISTS } from '../../graphQL/Queries';
import ListOfTasks from './ListOfTasks';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import { NEW_LIST_MUTATION } from '../../graphQL/Mutations';

const SideBar = () => {
    
    const [makeNewList, { errorNewList }] = useMutation(NEW_LIST_MUTATION);
    //todo: try and change 'refetch', to 'cache'
    const {error, loading, data, refetch} = useQuery(GET_ALL_LISTS);

    const handleNewList = (values) => {
        makeNewList({variables: values })
        .then(data => {
            if (!data.data) {
            console.log('something went wrong');
            }
            else
            console.log(data.data);
        });
        refetch();
        values.listName = '';
    
      }

    if(errorNewList){
        return <div>error...</div>
    }
    if(loading){
        return <div>loading...</div>
    }
    if(error){
        return <div>error...</div>
    }

    const validateNewList = Yup.object({
        listName: Yup.string()
          .required('A Name is Required'),
      })

    return (
        <Wrapper>
            <SideBar2>
                <H2><Img src={menu} alt="" /></H2>
                <Ul>
                    <Li><P>My day</P></Li>
                    <Li><P>Important</P></Li>
                    <Li><P>Planned</P></Li>
                    <Li><P>Assigned to you</P></Li>
                    <Li><P>Flagged email</P></Li>
                    <Li><P>Tasks</P></Li>
                </Ul>

                <ListOfTasks lists={data.lists} />
                
                <Formik
                    initialValues={{ listName: '' }}
                    validationSchema={validateNewList}
                    onSubmit={(handleNewList)}
                    >
                    {({values, errors}) => (
                        <Form>
                        <Field placeholder=" +  New List" autoComplete="off" name="listName" as={Input} /><br/>
                        </Form>
                    )}
                </Formik>
                
            </SideBar2>
        </Wrapper>
        );
    }
    
    export default SideBar;