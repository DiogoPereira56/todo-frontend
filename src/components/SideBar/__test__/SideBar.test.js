/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
import { NEW_LIST_MUTATION } from '../../../graphQL/Mutations';
import SideBar from '../SideBar';
import ReactDOM from 'react-dom';
import { fireEvent, prettyDOM, render, screen, waitFor, waitForElement } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';

const mockGetClientData = () => {
    return {
        idClient: 3,
        list: [
            { idList: 1, listName: 'lista', idClient: 3 },
            { idList: 2, listName: 'compras', idClient: 3 },
        ],
    };
};

const mocks = [
    {
        request: {
            query: NEW_LIST_MUTATION,
            variables: { listName: 'newList' },
        },
        result: {
            data: {
                addList: { idList: 331, idClient: 3, listName: 'newList' },
            },
        },
    },
];

describe('<SideBar />', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <MockedProvider>
                <SideBar></SideBar>
            </MockedProvider>,
            div,
        );
    });

    it('matches SideBar snapshot', () => {
        const mockedClientData = mockGetClientData();
        const tree = renderer
            .create(
                <MockedProvider>
                    <SideBar dataClient={mockedClientData}></SideBar>
                </MockedProvider>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders SideBar and its components correctly', async () => {
        const mockedClientData = mockGetClientData();
        const { getByTestId } = render(
            <MockedProvider mocks={[]} addTypename={false}>
                <SideBar dataClient={mockedClientData}></SideBar>
            </MockedProvider>,
        );
        //screen.debug();
        expect(getByTestId('lists')).toHaveTextContent('lista');
    });

    it('adds a new list correctly', async () => {
        const mockedClientData = mockGetClientData();
        const mockMutation = mocks;
        const { getByTestId } = render(
            <MockedProvider mocks={mockMutation} addTypename={false}>
                <SideBar dataClient={mockedClientData}></SideBar>
            </MockedProvider>,
        );
        let newListInput = getByTestId('NLInput');

        fireEvent.change(newListInput, { target: { value: 'newList' } });
        fireEvent.submit(newListInput);

        const loading = await waitFor(() => getByTestId('NLloading'));
        expect(loading).toHaveTextContent('Loading...');
    });
});
