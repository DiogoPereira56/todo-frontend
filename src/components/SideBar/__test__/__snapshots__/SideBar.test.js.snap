// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideBar /> matches SideBar snapshot 1`] = `
<div
  className="sc-bdfBwQ hNAEzB"
>
  <div
    className="sc-gsTCUz bAvXS"
  >
    <h2
      className="sc-dlfnbm ifTlkP"
    >
      <img
        alt=""
        className="sc-hKgILt cIkhnL"
        onClick={[Function]}
        src="menu.png"
      />
    </h2>
    <ul
      className="sc-iBPRYJ hMMeJ"
    >
      <li
        className="sc-eCssSg hfHFtt"
        onClick={[Function]}
      >
        <p
          className="sc-jSgupP bepHJl"
        >
          All Tasks
        </p>
      </li>
    </ul>
    <div
      data-testid="lists"
    >
      <li
        className="sc-eCssSg hfHFtt"
        onClick={[Function]}
      >
        <p
          className="sc-jSgupP bepHJl"
        >
          lista
        </p>
      </li>
      <li
        className="sc-eCssSg hfHFtt"
        onClick={[Function]}
      >
        <p
          className="sc-jSgupP bepHJl"
        >
          compras
        </p>
      </li>
    </div>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="sc-gKsewC eaSpDh"
        data-testid="NLInput"
        name="listName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=" +  New List"
        value=""
      />
      <br />
    </form>
    <div
      className="sc-fubCfw jUWFZp"
    >
      <div
        className="sc-pFZIQ iVNzqE"
      >
        1
      </div>
    </div>
  </div>
</div>
`;
